### PR TITLE
Split analysis of function definition and function body to ensure order of function definitions doesn't matter

### PR DESCRIPTION
**Pull request by cburgdorf**
### What was wrong?

As explained in #90 we do currently have a bug that causes us to fail to compile contracts with functions that call other functions if the called function isn't defined above the calling function. In plain english: The order of function definitions matters but it should not matter.

### How was it fixed?

1. `func_def` was split into two functions: `func_def` and `func_body`. The former is only concerned about checking the types of the function definition and creating a scope for the function. It does not perform any tasks on the `body`  of the function.

2. The `scope` of the function is now also preserved when calling `add_function` on the contract scope. This is to give `function_body` a clean way to resolve the (already existing) scope.

3. In `contract_def`, we add a second pass that iterates over all functions again (after they have been iterated to call `func_def`) to call `func_body`. This is effectively what causes the order of the function definitions to become insignificant.

4. `func_body` retrieves the function scope by looking up the function definition in the contract scope and performs all checks on the function body.

### CODE REVIEW

File: semantics/src/traversal/functions.rs [hunk 1]
```diff
@@ -28,8 +28,6 @@ use fe_parser::ast as fe;
 use fe_parser::span::Spanned;
 use std::rc::Rc;
 
-/// Gather context information for a function definition and check for type
```
Comment (g-r-a-n-t) [inline semantics/src/traversal/functions.rs:31]: We should keep doc comments on all public functions

File: semantics/src/traversal/functions.rs [hunk 2]
```diff
@@ -82,7 +72,44 @@ pub fn func_def(
 
         context.borrow_mut().add_function(def, attributes);
 
-        traverse_statements(function_scope, context, body)?;
+        return Ok(());
+    }
+
+    unreachable!();
+}
+
```
Comment (g-r-a-n-t) [inline semantics/src/traversal/functions.rs:80]: same, doc comment here too

**Review by g-r-a-n-t — APPROVED** [review]
Looks good :+1: just a few comments with regard to documentation

File: semantics/src/traversal/functions.rs (hunk 1 above)
Comment (cburgdorf) [reply semantics/src/traversal/functions.rs:31]: Good point. Removal was by accident.
